{% extends "base.html" %} {% block content %}
<div class="main-content">
  <div class="container mt-5">
    <h1 class="text-center padding">Upload Datasets</h1>
    <div class="line"></div>
    <!-- Error Messages -->
    <div id="dynamic-error-section"></div>

    <!-- Upload Alert -->
    {% if messages %} {% for message in messages %}
    <div class="alert alert-secondary alert-dismissible fade show" role="alert">
      <div class="text-center">
        <i class="bi bi-check-circle-fill"></i>
        <strong>{{ message }}</strong>
      </div>
      <button
        type="button"
        class="btn-close"
        data-bs-dismiss="alert"
        aria-label="Close"
      ></button>
    </div>
    {% endfor %} {% endif %}

    <!-- Upload Section -->
    <form method="post" enctype="multipart/form-data" class="container mb-5">
      {% csrf_token %}
      <div class="row">
        <!-- Dataset Folder Upload -->
        <div class="col-md-6">
          <div class="drag-area" id="datasetFileInput">
            <i class="bi bi-download" aria-hidden="true"></i>
            <p>Drag and Drop <b>Dataset Folder</b> Here or Browse</p>
            <input
              type="file"
              class="file-input"
              id="file-input"
              name="dataset-file"
              onchange="updateFileInfo(this, 'zip-info')"
              webkitdirectory
              directory
              multiple
              required
            />
            <input type="text" id="directories" name="directories" hidden />
            <div class="upload-info" id="zip-info"></div>
          </div>
        </div>
        <!-- Metadata JSON File Upload -->
        <div class="col-md-6">
          <div class="drag-area" id="jsonFileInput">
            <i class="bi bi-download" aria-hidden="true"></i>
            <p>Drag and Drop <b>Metadata JSON File</b> Here or Browse</p>
            <input
              type="file"
              class="file-input"
              accept=".json"
              onchange="readJSON(event)"
              id="json-file"
              name="json-file"
              required
            />
            <div class="upload-info" id="json-info"></div>
          </div>
        </div>
        <button type="submit" class="btn btn-submit mt-3">Submit</button>
      </div>
    </form>

    <!-- Instruction Box -->
    <div id="uploadInstructions" class="bg-light p-5 rounded">
      <h2 class="text-center">Upload Instructions</h2>
      <div class="row">
        <div class="col-4">
          <div
            id="instructions-navbar"
            class="nav-pills flex-column sticky-top"
            style="top: 20px"
          >
            <a class="nav-link" href="#item-1">Upload Boxes</a>
            <a class="nav-link ms-3 my-1" href="#item-1-1"
              >Dataset Folder Upload</a
            >
            <a class="nav-link ms-3 my-1" href="#item-1-2"
              >Metadata JSON Upload</a
            >
            <a class="nav-link ms-3 my-1" href="#item-1-3">General Usage</a>
            <a class="nav-link" href="#item-2">Metadata Card</a>
            <a class="nav-link ms-3 my-1" href="#item-2-1">Overview</a>
          </div>
        </div>
        <div class="col-8">
          <div
            data-bs-spy="scroll"
            data-bs-target="#instructions-navbar"
            data-bs-smooth-scroll="true"
            class="uploadInstructions"
            tabindex="0"
            style="height: 800px; overflow-y: auto"
          >
            <h3 id="item-1">Upload Boxes</h3>
            <h5 id="item-1-1">Dataset Folder Upload</h5>
            <p>
              For the dataset upload, please select the folder containing your
              dataset files. For single files please also upload them in a
              folder.
              <br />
              <br />
              Please note that the <b>Maximum Number of Files is 1000</b> and the <b>Maximum Dataset Size is 1GB</b>.
            </p>
            <h5 id="item-1-2">Metadata JSON Upload</h5>
            <p>
              You are required to fill out a metadata card for your dataset.
              This includes providing information such as the dataset title,
              description, keywords, and other relevant details that will be
              covered below. Please download the metadata template, fill it out,
              and upload it along with your dataset.
            </p>
            <h5 id="item-1-3">General Usage</h5>
            <p>
              When uploading files, you can either drag and drop the files into
              the box or click on the box to browse your files. If you wish to
              upload a different file, you can click on the box again or drag
              and drop to select a new file.
            </p>
            <div class="text-center">
              <a
                href="/static/metadata_standard/metadata_v5.json"
                download="metadata_template.json"
                class="btn btn-submit"
                >Download Metadata Card</a
              >
            </div>
            <h3 id="item-2">Metadata Card</h3>
            <h5 id="item-2-1">Overview</h5>
            <p>
              The metadata card is a JSON file that contains various fields to
              describe the nature of your dataset. Some of these fields are
              mandatory while others are optional. Below is a brief overview of
              the different fields and whether they are Mandatory or Optional:
            </p>
            <ul class="metadata-list">
              <!-- Basic Identity -->
              <li onclick="toggleVisibility(event, 'basic_identity')">
                <span class="toggle-symbol">+</span>Basic Identity
                <ul id="basic_identity" class="nested">
                  <li>- Title: Mandatory</li>
                  <li>- Description: Mandatory</li>
                  <li>- Version: Mandatory</li>
                  <li>- Changelog: Optional</li>
                  <li>- Keywords: Mandatory</li>
                  <li>- License: Optional</li>
                </ul>
              </li>
              <!-- Author Contact Details -->
              <li onclick="toggleVisibility(event, 'author_contact_details')">
                <span class="toggle-symbol">+</span>Author Contact Details
                <ul id="author_contact_details" class="nested">
                  <li>- Affiliation: Optional</li>
                  <li>- Email: Mandatory</li>
                  <li>- Phone: Optional</li>
                </ul>
              </li>
              <!-- Dataset Creation -->
              <li onclick="toggleVisibility(event, 'dataset_creation')">
                <span class="toggle-symbol">+</span>Dataset Creation
                <ul id="dataset_creation" class="nested">
                  <li onclick="toggleVisibility(event, 'general_creation')">
                    <span class="toggle-symbol">+</span>General
                    <ul id="general_creation" class="nested">
                      <li>- Data Origin: Mandatory</li>
                      <li onclick="toggleVisibility(event, 'synthetic_details')">
                        <span class="toggle-symbol">+</span>Synthetic Data Details
                        <ul id="synthetic_details" class="nested">
                          <li>- Generation Method: Mandatory</li>
                          <li>- Software Used: Optional</li>
                        </ul>
                      </li>
                      <li onclick="toggleVisibility(event, 'natural_details')">
                        <span class="toggle-symbol">+</span>Natural Data Details
                        <ul id="natural_details" class="nested">
                          <li>- Collection Circumstances: Mandatory</li>
                          <li>- Collection Date: Optional</li>
                          <li>- Collection Location: Optional</li>
                        </ul>
                      </li>
                    </ul>
                  </li>
                  <li onclick="toggleVisibility(event, 'data_completion')">
                    <span class="toggle-symbol">+</span>Data Completion
                    <ul id="data_completion" class="nested">
                      <li>- Dataset Status: Mandatory</li>
                      <li>- Future Work: Optional</li>
                    </ul>
                  </li>
                  <li onclick="toggleVisibility(event, 'pre_processing')">
                    <span class="toggle-symbol">+</span>Pre Processing
                    <ul id="pre_processing" class="nested">
                      <li onclick="toggleVisibility(event, 'data_cleanliness')">
                        <span class="toggle-symbol">+</span>Data Cleanliness
                        <ul id="data_cleanliness" class="nested">
                          <li>- Cleanliness Status: Mandatory</li>
                          <li>- Cleaning Details: Optional</li>
                        </ul>
                      </li>
                      <li onclick="toggleVisibility(event, 'labeling')">
                        <span class="toggle-symbol">+</span>Labeling
                        <ul id="labeling" class="nested">
                          <li>- Labeled: Mandatory</li>
                          <li>- Labeling Details: Optional</li>
                        </ul>
                      </li>
                      <li onclick="toggleVisibility(event, 'data_split')">
                        <span class="toggle-symbol">+</span>Data Split
                        <ul id="data_split" class="nested">
                          <li>- Train Split: Optional</li>
                          <li>- Validation Split: Optional</li>
                          <li>- Test Split: Optional</li>
                        </ul>
                      </li>
                    </ul>
                  </li>
                </ul>
              </li>
              <!-- Dataset Usage -->
              <li onclick="toggleVisibility(event, 'dataset_usage')">
                <span class="toggle-symbol">+</span>Dataset Usage
                <ul id="dataset_usage" class="nested">
                  <li>- Citation: Optional</li>
                  <li>- Data Usage Description: Mandatory</li>
                  <li>- Related Datasets: Optional</li>
                  <li>- Related Publications: Optional</li>
                </ul>
              </li>
              <!-- Dataset Composition -->
              <li onclick="toggleVisibility(event, 'dataset_composition')">
                <span class="toggle-symbol">+</span>Dataset Composition
                <ul id="dataset_composition" class="nested">
                  <li onclick="toggleVisibility(event, 'general_composition')">
                    <span class="toggle-symbol">+</span>General
                    <ul id="general_composition" the "nested">
                      <li>- Format: Mandatory</li>
                      <li>- Dataset Size: Mandatory</li>
                      <li>- Number of Files: Mandatory</li>
                      <li>- Average File Size: Mandatory</li>
                    </ul>
                  </li>
                  <li onclick="toggleVisibility(event, 'csv_details')">
                    <span class="toggle-symbol">+</span>CSV
                    <ul id="csv_details" class="nested">
                      <li>- Delimiter: Mandatory</li>
                      <li>- Has Header: Mandatory</li>
                      <li>- Column Names: Mandatory</li>
                    </ul>
                  </li>
                  <li onclick="toggleVisibility(event, 'excel_details')">
                    <span class="toggle-symbol">+</span>Excel
                    <ul id="excel_details" class="nested">
                      <li>- Has Macros: Mandatory</li>
                      <li>- Average Sheets: Optional</li>
                    </ul>
                  </li>
                  <li onclick="toggleVisibility(event, 'image_details')">
                    <span class="toggle-symbol">+</span>Image
                    <ul id="image_details" class="nested">
                      <li>- File Type: Mandatory</li>
                      <li>- Average Resolution: Mandatory</li>
                      <li>- Color Mode: Mandatory</li>
                      <li>- Image Processing: Optional</li>
                    </ul>
                  </li>
                  <li onclick="toggleVisibility(event, 'video_details')">
                    <span class="toggle-symbol">+</span>Video
                    <ul id="video_details" class="nested">
                      <li>- File Type: Mandatory</li>
                      <li>- Average Duration: Mandatory</li>
                      <li>- Video Resolution: Mandatory</li>
                      <li>- Video Processing: Optional</li>
                    </ul>
                  </li>
                  <li onclick="toggleVisibility(event, 'audio_details')">
                    <span class="toggle-symbol">+</span>Audio
                    <ul id="audio_details" class="nested">
                      <li>- File Type: Mandatory</li>
                      <li>- Average Duration: Mandatory</li>
                      <li>- Average Bitrate: Optional</li>
                      <li>- Audio Processing: Optional</li>
                    </ul>
                  </li>
                  <li onclick="toggleVisibility(event, 'fasta_details')">
                    <span class="toggle-symbol">+</span>FASTA
                    <ul id="fasta_details" class="nested">
                      <li>- Sequence Type: Mandatory</li>
                      <li>- Organism Group: Mandatory</li>
                      <li>- Source Organism Detail: Optional</li>
                      <li>- Minimum Sequence Length: Mandatory</li>
                      <li>- Maximum Sequence Length: Mandatory</li>
                    </ul>
                  </li>
                  <li onclick="toggleVisibility(event, 'genbank_details')">
                    <span class="toggle-symbol">+</span>GenBank
                    <ul id="genbank_details" class="nested">
                      <li>- Sequence Type: Mandatory</li>
                      <li>- Organism Group: Mandatory</li>
                      <li>- Source Organism Detail: Optional</li>
                      <li>- Minimum Sequence Length: Mandatory</li>
                      <li>- Maximum Sequence Length: Mandatory</li>
                    </ul>
                  </li>
                  <li onclick="toggleVisibility(event, 'text_details')">
                    <span class="toggle-symbol">+</span>Text
                    <ul id="text_details" class="nested">
                      <li>- Encoding: Mandatory</li>
                      <li>- Language: Mandatory</li>
                      <li>- Text Format: Optional</li>
                    </ul>
                  </li>
                  <li onclick="toggleVisibility(event, 'pdf_details')">
                    <span class="toggle-symbol">+</span>PDF
                    <ul id="pdf_details" class="nested">
                      <li>- Text Extractable: Mandatory</li>
                      <li>- PDF Version: Optional</li>
                    </ul>
                  </li>
                </ul>
              </li>
              <!-- Ethics -->
              <li onclick="toggleVisibility(event, 'ethics')">
                <span class="toggle-symbol">+</span>Ethics
                <ul id="ethics" class="nested">
                  <li>- Data Collection Ethics: Optional</li>
                  <li>- Data Usage Ethics: Optional</li>
                  <li>- Ethical Review Status: Optional</li>
                </ul>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>

    <!-- File Hierarchy Storage Script -->
    <script>
      document
        .querySelector("#file-input")
        .addEventListener("change", function () {
          var files = document.querySelector("#file-input").files;
          var directories = {};
          for (var file of files) {
            directories[file.name] = file.webkitRelativePath;
          }
          directories = JSON.stringify(directories);
          console.log(directories);

          document.querySelector("#directories").value = directories;
          validateFiles(files);
        });

      function validateFiles(files) {
        var totalSize = 0;
        var errorSection = document.getElementById("dynamic-error-section");
        errorSection.innerHTML = ""; // Clear previous errors.

        if (files.length > 1000) {
          var errorMessage = "Sorry, you cannot upload more than 1000 files.";
          errorSection.innerHTML += `
              <div class="alert alert-danger alert-dismissible fade show" role="alert">
                <div class="text-center">
                  <i class="bi bi-exclamation-triangle-fill"></i>
                  <strong>${errorMessage}</strong>
                  <button
                    type="button"
                    class="btn-close"
                    data-bs-dismiss="alert"
                    aria-label="Close"
                  ></button>
                </div>
              </div>
            `;
          document.querySelector("#file-input").value = ""; // Clear the selected files.
          return false;
        }

        for (var i = 0; i < files.length; i++) {
          totalSize += files[i].size;
        }

        if (totalSize > 1073741824) {
          // 1GB in bytes
          var errorMessage = "Sorry, the total dataset size cannot exceed 1GB.";
          errorSection.innerHTML += `
              <div class="alert alert-danger alert-dismissible fade show" role="alert">
                <div class="text-center">
                  <i class="bi bi-exclamation-triangle-fill"></i>
                  <strong>${errorMessage}</strong>
                  <button
                    type="button"
                    class="btn-close"
                    data-bs-dismiss="alert"
                    aria-label="Close"
                  ></button>
                </div>
              </div>
            `;
          document.querySelector("#file-input").value = ""; // Clear the selected files.
          return false;
        }
      }
    </script>

    <!-- JSON Validation Script -->
    <script>
      // JSON Validation Script
      function readJSON() {
        const fileInput = document.getElementById("json-file");
        const file = fileInput.files[0];

        const reader = new FileReader();

        reader.onload = function (event) {
          const fileContent = event.target.result;

          // Clear previous errors.
          document.getElementById("dynamic-error-section").innerHTML = "";

          try {
            const jsonObj = JSON.parse(fileContent);

            if (validateJSON(jsonObj)) {
              updateFileInfo(fileInput, "json-info");
              return true;
            } else {
              return false;
            }
          } catch (e) {
            console.log(e);
            document.getElementById("dynamic-error-section").insertAdjacentHTML(
              "beforeend",
              `
              <div class="alert alert-danger alert-dismissible fade show" role="alert">
                <div class="text-center">
                <i class="bi bi-exclamation-triangle-fill"></i>
                  <strong>Could not parse the JSON File, please ensure the format is correct.</strong>
                  <button type="button" class="btn-close" data-bs-dismiss="alert" aria-label="Close"></button>
                </div>
              </div>
            `
            );

            return false;
          }
        };

        reader.readAsText(file);
      }

      function validateJSON(obj) {
        let errors = [];

        function isValidString(value, placeholder) {
          return value && value !== placeholder;
        }

        function isValidFloat(value) {
          return !isNaN(value) && (Number(value) === value);
        }

        function isValidInteger(value) {
          return !isNaN(value) && Number.isInteger(Number(value));
        }

        function isValidDate(value) {
          return !isNaN(Date.parse(value));
        }

        function isValidKeywords(value) {
          if (Array.isArray(value)) {
            return value.length >= 3;
          } else if (typeof value === 'string') {
            return value.split(/[,;|]/).map(s => s.trim()).filter(Boolean).length >= 3;
          }
          return false;
        }

        function isValidCategory(value, categories) {
          return categories.map(c => c.toLowerCase()).includes(value.toLowerCase());
        }

        function isValidEmail(value) {
          const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
          return emailRegex.test(value);
        }

        function isValidPercentage(value) {
          return isValidFloat(value);
        }

        {% comment %} // Validate Basic Identity
        if (!isValidString(obj.basic_identity?.title, "Mandatory - Title of the Dataset.")) {
          errors.push("Your dataset metadata is missing the Title or it is not properly filled.");
        }
        if (!isValidString(obj.basic_identity?.description, "Mandatory - A detailed description of the dataset and its contents.")) {
          errors.push("Your dataset metadata is missing the Description or it is not properly filled.");
        }
        if (!isValidFloat(Number(obj.basic_identity?.version))) {
          errors.push("Your dataset metadata is missing the Version or it is not a float value.");
        }
        if (!isValidKeywords(obj.basic_identity?.keywords)) {
          errors.push("Your dataset metadata is missing the Keywords or there are less than 3 keywords in the correct format.");
        }
        if (!isValidString(obj.basic_identity?.license, "Mandatory - Type of license under which the dataset is distributed.")) {
          errors.push("Your dataset metadata is missing the License or it is not properly filled.");
        }

        // Validate Author Contact Details
        if (!isValidEmail(obj.author_contact_details?.email)) {
          errors.push("Your dataset metadata is missing the Author Email or it is not a valid email address.");
        }

        // Validate Dataset Creation
        if (!isValidString(obj.dataset_creation?.general?.data_origin, "Mandatory - Indicates whether the data is ('Synthetically-Generated'/ 'Naturally-Obtained'/ 'Other').")) {
          errors.push("Your dataset metadata is missing the Data Origin or it is not properly filled.");
        } else {
          const origin = obj.dataset_creation.general.data_origin.toLowerCase();
          if (origin.includes("synthetically-generated")) {
            if (!isValidString(obj.dataset_creation.general.synthetic_data_details?.generation_method, "Mandatory - Method used for generating synthetic data, if applicable.")) {
              errors.push("Your dataset metadata is missing the Synthetic Data Generation Method or it is not properly filled.");
            }
            if (!isValidString(obj.dataset_creation.general.synthetic_data_details?.software_used, "Mandatory - Details of any software or tools used for generating synthetic data.")) {
              errors.push("Your dataset metadata is missing the Synthetic Data Software Used or it is not properly filled.");
            }
          }
          if (origin.includes("naturally-obtained")) {
            const studyTypes = ['Observational', 'Experimental', 'Clinical Trial', 'Field Study', 'Laboratory Study', 'Survey Research', 'Longitudinal Study', 'Cross-Sectional Study', 'Case Study', 'Other'];
            if (!isValidCategory(obj.dataset_creation.general.natural_data_details?.study_type, studyTypes)) {
              errors.push("Your dataset metadata is missing the Natural Data Study Type or it is not one of the specified categories.");
            }
            if (!isValidString(obj.dataset_creation.general.natural_data_details?.experimental_setup?.general_description, "Mandatory - General description of the study, outlining its main objectives and methodology.")) {
              errors.push("Your dataset metadata is missing the Natural Data Experimental Setup Description or it is not properly filled.");
            }
            if (!isValidDate(obj.dataset_creation.general.natural_data_details?.experimental_setup?.study_start_date)) {
              errors.push("Your dataset metadata is missing the Natural Data Study Start Date or it is not in the correct format (mm/dd/yyyy).");
            }
            if (!isValidDate(obj.dataset_creation.general.natural_data_details?.experimental_setup?.study_end_date)) {
              errors.push("Your dataset metadata is missing the Natural Data Study End Date or it is not in the correct format (mm/dd/yyyy).");
            }
            if (!isValidString(obj.dataset_creation.general.natural_data_details?.experimental_conditions?.experiment_location, "Mandatory - Location where the experiment was conducted, including specific site details and environmental conditions, if relevant.")) {
              errors.push("Your dataset metadata is missing the Natural Data Experiment Location or it is not properly filled.");
            }
          }
        }

        // Validate Data Completion
        const datasetStatuses = ['Complete', 'In Progress', 'Abandoned & Incomplete', 'Other'];
        if (!isValidCategory(obj.dataset_creation?.data_completion?.dataset_status, datasetStatuses)) {
          errors.push("Your dataset metadata is missing the Dataset Status or it is not one of the specified categories.");
        }

        // Validate Pre Processing
        const yesNoNA = ['Yes', 'No', 'N/A'];
        if (!isValidCategory(obj.dataset_creation?.pre_processing?.raw_data, yesNoNA)) {
          errors.push("Your dataset metadata is missing the Raw Data field or it is not 'Yes'/'No'.");
        }
        const cleanlinessStatuses = ['Clean', 'Partially Clean', 'Unclean', 'N/A'];
        if (!isValidCategory(obj.dataset_creation?.pre_processing?.data_cleanliness?.cleanliness_status, cleanlinessStatuses)) {
          errors.push("Your dataset metadata is missing the Data Cleanliness Status or it is not one of the specified categories.");
        }
        if (!isValidCategory(obj.dataset_creation?.pre_processing?.labeling?.labeled, yesNoNA)) {
          errors.push("Your dataset metadata is missing the Labeling field or it is not 'Yes'/'No'.");
        }
        if (obj.dataset_creation?.pre_processing?.data_split?.train_split && !isValidPercentage(obj.dataset_creation.pre_processing.data_split.train_split)) {
          errors.push("Train split must be a number (integer or float).");
        }
        if (obj.dataset_creation?.pre_processing?.data_split?.validation_split && !isValidPercentage(obj.dataset_creation.pre_processing.data_split.validation_split)) {
          errors.push("Validation split must be a number (integer or float).");
        }
        if (obj.dataset_creation?.pre_processing?.data_split?.test_split && !isValidPercentage(obj.dataset_creation.pre_processing.data_split.test_split)) {
          errors.push("Test split must be a number (integer or float).");
        }

        // Validate Dataset Usage
        if (!isValidString(obj.dataset_usage?.data_usage_description, "Mandatory - Description of how the data can be used, including any restrictions or guidelines.")) {
          errors.push("Your dataset metadata is missing the Data Usage Description or it is not properly filled.");
        }

        // Validate Dataset Composition
        if (!isValidString(obj.dataset_composition?.general?.format, "Mandatory - File format ('CSV'/ 'Excel'/ 'Image'/ 'Video'/ 'Audio'/ 'FASTA'/ 'GenBank'/ 'Text'/ 'PDF'/ 'Matlab'/ 'Other').")) {
          errors.push("Your dataset metadata is missing the File Format or it is not properly filled.");
        } else {
          const format = obj.dataset_composition.general.format.toLowerCase();
          if (format.includes("csv")) {
            if (!isValidString(obj.dataset_composition.csv?.delimiter, "Mandatory - Delimiter used in CSV files ('Comma'/ 'Colon'/ 'Semicolon'/ 'Other').")) {
              errors.push("Your dataset metadata is missing the CSV Delimiter or it is not properly filled.");
            }
            if (!isValidCategory(obj.dataset_composition.csv?.has_header, yesNoNA)) {
              errors.push("Your dataset metadata is missing the CSV Has Header field or it is not 'Yes'/'No'.");
            }
            if (!Array.isArray(obj.dataset_composition.csv?.column_names) || obj.dataset_composition.csv.column_names.length === 0) {
              errors.push("Your dataset metadata is missing the CSV Column Names or it is not a non-empty array.");
            }
          }
          if (format.includes("excel")) {
            if (!isValidCategory(obj.dataset_composition.excel?.has_macros, yesNoNA)) {
              errors.push("Your dataset metadata is missing the Excel Has Macros field or it is not 'Yes'/'No'.");
            }
            if (!isValidInteger(Number(obj.dataset_composition.excel?.average_sheets))) {
              errors.push("Your dataset metadata is missing the Excel Average Sheets or it is not an integer.");
            }
          }
          if (format.includes("image")) {
            if (!isValidString(obj.dataset_composition.image?.file_type, "Mandatory - File type of images ('JPEG'/ 'PNG'/ 'GIF'/ 'TIFF'/ 'BMP'/ 'Other').")) {
              errors.push("Your dataset metadata is missing the Image File Type or it is not properly filled.");
            }
            if (!isValidFloat(Number(obj.dataset_composition.image?.average_resolution))) {
              errors.push("Your dataset metadata is missing the Image Average Resolution or it is not a float value.");
            }
            const colorModes = ['RGB', 'CMYK', 'Grayscale', 'Other'];
            if (!isValidCategory(obj.dataset_composition.image?.color_mode, colorModes)) {
              errors.push("Your dataset metadata is missing the Image Color Mode or it is not one of the specified categories.");
            }
          }
          if (format.includes("video")) {
            if (!isValidString(obj.dataset_composition.video?.file_type, "Mandatory - File type of videos ('MP4'/ 'AVI'/ 'MOV'/ 'WMV'/ 'MKV'/ 'Other').")) {
              errors.push("Your dataset metadata is missing the Video File Type or it is not properly filled.");
            }
            if (!isValidFloat(Number(obj.dataset_composition.video?.average_duration))) {
              errors.push("Your dataset metadata is missing the Video Average Duration or it is not a float value.");
            }
            const videoResolutions = ['Lower than SD', 'SD', 'HD', 'Full HD', '2K', '4K', '8K', 'Greater than 8K'];
            if (!isValidCategory(obj.dataset_composition.video?.video_resolution, videoResolutions)) {
              errors.push("Your dataset metadata is missing the Video Resolution or it is not one of the specified categories.");
            }
          }
          if (format.includes("audio")) {
            if (!isValidString(obj.dataset_composition.audio?.file_type, "Mandatory - File type of audio files ('MP3'/ 'WAV'/ 'AAC'/ 'FLAC'/ 'OGG'/ 'Other').")) {
              errors.push("Your dataset metadata is missing the Audio File Type or it is not properly filled.");
            }
            if (!isValidFloat(Number(obj.dataset_composition.audio?.average_duration))) {
              errors.push("Your dataset metadata is missing the Audio Average Duration or it is not a float value.");
            }
          }
          if (format.includes("fasta")) {
            if (!isValidString(obj.dataset_composition.fasta?.general_info?.sequence_type, "Mandatory - List of predominant sequence types in the FASTA files ('DNA'/ 'RNA'/ 'cDNA'/ 'Amino Acid'/ 'Other').")) {
              errors.push("Your dataset metadata is missing the FASTA Sequence Type or it is not properly filled.");
            }
            const organismGroups = ['Bacteria', 'Virus', 'Eukaryote', 'Archaea', 'Organelles', 'Other'];
            if (!isValidCategory(obj.dataset_composition.fasta?.general_info?.organism_group, organismGroups)) {
              errors.push("Your dataset metadata is missing the FASTA Organism Group or it is not one of the specified categories.");
            }
            if (!isValidInteger(Number(obj.dataset_composition.fasta?.general_info?.minimum_sequence_length))) {
              errors.push("Your dataset metadata is missing the FASTA Minimum Sequence Length or it is not an integer.");
            }
            if (!isValidInteger(Number(obj.dataset_composition.fasta?.general_info?.maximum_sequence_length))) {
              errors.push("Your dataset metadata is missing the FASTA Maximum Sequence Length or it is not an integer.");
            }
            const sequencingMethods = ['Illumina', 'PacBio SMRT', 'Nanopore Sequencing', 'Sanger', 'Ion Torrent', 'Pyrosequencing', 'SOLiD Sequencing', 'Other'];
            if (!isValidCategory(obj.dataset_composition.fasta?.sequencing_info?.sequencing_method, sequencingMethods)) {
              errors.push("Your dataset metadata is missing the FASTA Sequencing Method or it is not one of the specified categories.");
            }
          }
          if (format.includes("genbank")) {
            if (!isValidString(obj.dataset_composition.genbank?.general_info?.sequence_type, "Mandatory - List of predominant sequence types in the GenBank files ('DNA'/ 'RNA'/ 'cDNA'/ 'Amino Acid'/ 'Other').")) {
              errors.push("Your dataset metadata is missing the GenBank Sequence Type or it is not properly filled.");
            }
            if (!isValidCategory(obj.dataset_composition.genbank?.general_info?.organism_group, organismGroups)) {
              errors.push("Your dataset metadata is missing the GenBank Organism Group or it is not one of the specified categories.");
            }
            if (!isValidInteger(Number(obj.dataset_composition.genbank?.general_info?.minimum_sequence_length))) {
              errors.push("Your dataset metadata is missing the GenBank Minimum Sequence Length or it is not an integer.");
            }
            if (!isValidInteger(Number(obj.dataset_composition.genbank?.general_info?.maximum_sequence_length))) {
              errors.push("Your dataset metadata is missing the GenBank Maximum Sequence Length or it is not an integer.");
            }
            if (!isValidCategory(obj.dataset_composition.genbank?.sequencing_info?.sequencing_method, sequencingMethods)) {
              errors.push("Your dataset metadata is missing the GenBank Sequencing Method or it is not one of the specified categories.");
            }
          }
          if (format.includes("matlab")) {
            if (!isValidString(obj.dataset_composition.matlab?.matlab_version, "Mandatory - Version of MATLAB used to generate the data (e.g., R2019b).")) {
              errors.push("Your dataset metadata is missing the Matlab Version or it is not properly filled.");
            }
            const matlabVariables = ['Matrix', 'Cell Array', 'Structure', 'Table', 'Function Handle', 'Time Series', 'Other'];
            if (!isValidCategory(obj.dataset_composition.matlab?.variables_stored, matlabVariables)) {
              errors.push("Your dataset metadata is missing the Matlab Variables Stored or it is not one of the specified categories.");
            }
          }
          if (format.includes("text")) {
            if (!isValidString(obj.dataset_composition.text?.encoding, "Mandatory - Character encoding used in text files ('UTF-8'/ 'UTF-16'/ 'UTF-32'/ 'ASCII'/ 'Other').")) {
              errors.push("Your dataset metadata is missing the Text Encoding or it is not properly filled.");
            }
            if (!isValidString(obj.dataset_composition.text?.language, "Mandatory - Predominant language of text files (e.g., English, Spanish, Chinese).")) {
              errors.push("Your dataset metadata is missing the Text Language or it is not properly filled.");
            }
          }
          if (format.includes("pdf")) {
            if (!isValidCategory(obj.dataset_composition.pdf?.text_extractable, [...yesNoNA, 'N/A'])) {
              errors.push("Your dataset metadata is missing the PDF Text Extractable field or it is not 'Yes'/'No'/'N/A'.");
            }
          }
        }

        // Validate optional fields only if they are filled out and not containing the default instructions
        function validateOptionalField(field, typeCheckFunction, errorMsg, defaultInstruction) {
          if (field !== undefined && field !== defaultInstruction && !typeCheckFunction(field)) {
            errors.push(errorMsg);
          }
        }

        validateOptionalField(obj.basic_identity?.changelog, isValidString, "Changelog must be a valid string.", "Optional - Description of changes made compared to the previous version of the dataset.");
        validateOptionalField(obj.author_contact_details?.affiliation, isValidString, "Affiliation must be a valid string.", "Optional - Affiliation of the author.");
        validateOptionalField(obj.author_contact_details?.phone, isValidString, "Phone must be a valid string.", "Optional - Phone number of the author.");
        validateOptionalField(obj.dataset_creation.general.synthetic_data_details?.related_publications, isValidString, "Related publications must be a valid string.", "Optional - Citations or links to publications that are directly related to the synthetic data generation.");
        validateOptionalField(obj.dataset_creation.general.natural_data_details?.experimental_setup?.instrumentation, isValidString, "Instrumentation must be a valid string.", "Optional - Specific instruments and technologies used to collect the data.");
        validateOptionalField(obj.dataset_creation.general.natural_data_details?.experimental_setup?.protocols, isValidString, "Protocols must be a valid string.", "Optional - List any common protocols used during the experiment.");
        validateOptionalField(obj.dataset_creation.general.natural_data_details?.sample_information?.number_of_samples, isValidInteger, "Number of samples must be an integer.", "Optional - Total number of samples collected and analyzed in the study.");
        validateOptionalField(obj.dataset_creation.general.natural_data_details?.sample_information?.sample_types, isValidString, "Sample types must be a valid string.", "Optional - Description of the types of samples collected, such as biological tissues, environmental specimens, clinical samples, etc., including specific details when applicable.");
        validateOptionalField(obj.dataset_creation.general.natural_data_details?.sample_information?.participants?.total_participants, isValidInteger, "Total participants must be an integer.", "Optional - Total number of participants involved in the study, if applicable.");
        validateOptionalField(obj.dataset_creation.general.natural_data_details?.sample_information?.participants?.demographic_data, isValidString, "Demographic data must be a valid string.", "Optional - Information about the demographic characteristics of the participants (age range, gender, ethnicity, etc.).");
        validateOptionalField(obj.dataset_creation.general.natural_data_details?.sample_information?.participants?.selection_criteria, isValidString, "Selection criteria must be a valid string.", "Optional - Criteria used for selecting participants, including any inclusion or exclusion criteria.");
        validateOptionalField(obj.dataset_creation.general.natural_data_details?.sample_information?.participants?.health_status, isValidString, "Health status must be a valid string.", "Optional - Health status or medical conditions of participants relevant to the study.");
        validateOptionalField(obj.dataset_creation.general.natural_data_details?.experimental_conditions?.experiment_conditions, isValidString, "Experimental conditions must be a valid string.", "Optional - Detailed conditions under which the experiment was performed, such as temperature, lighting, and other relevant environmental parameters.");
        validateOptionalField(obj.dataset_creation.general.natural_data_details?.related_publications, isValidString, "Related publications must be a valid string.", "Optional - Citations or links to publications that are directly related to the dataset's usage.");
        validateOptionalField(obj.dataset_creation.data_completion?.future_work, isValidString, "Future work must be a valid string.", "Optional - Description of any planned future work or improvements to be made on the dataset.");
        validateOptionalField(obj.dataset_creation.pre_processing?.data_cleanliness?.cleaning_details, isValidString, "Cleaning details must be a valid string.", "Optional - Description of the cleaning process applied to the dataset (e.g., methods used, types of errors corrected).");
        validateOptionalField(obj.dataset_creation.pre_processing?.labeling?.labeling_details, isValidString, "Labeling details must be a valid string.", "Optional - Description of the labeling process and the types of labels used.");
        validateOptionalField(obj.dataset_usage?.citation, isValidString, "Citation must be a valid string.", "Optional - Recommended citation for the dataset.");
        validateOptionalField(obj.dataset_usage?.related_datasets, isValidString, "Related datasets must be a valid string.", "Optional - Links to related datasets.");
        validateOptionalField(obj.dataset_usage?.related_publications, isValidString, "Related publications must be a valid string.", "Optional - Links or references to any papers related to the dataset's usage.");
        validateOptionalField(obj.dataset_composition.image?.image_processing, isValidString, "Image processing must be a valid string.", "Optional - Description of any image processing applied (e.g., normalization, augmentation).");
        validateOptionalField(obj.dataset_composition.video?.video_processing, isValidString, "Video processing must be a valid string.", "Optional - Description of any video processing applied (e.g., frame extraction, stabilization).");
        validateOptionalField(obj.dataset_composition.audio?.average_bitrate, isValidFloat, "Audio average bitrate must be a float value.", "Optional - Average bitrate across audio files in bits per second.");
        validateOptionalField(obj.dataset_composition.audio?.audio_processing, isValidString, "Audio processing must be a valid string.", "Optional - Description of any audio processing applied (e.g., noise reduction, normalization).");
        validateOptionalField(obj.dataset_composition.fasta?.general_info?.specific_species, isValidString, "FASTA specific species must be a valid string.", "Optional - List of the specific species or strains represented in the FASTA files.");
        validateOptionalField(obj.dataset_composition.fasta?.sequencing_info?.coverage_depth, isValidInteger, "FASTA coverage depth must be an integer.", "Optional - Average depth of coverage, crucial for assessing data quality as an Integer.");
        validateOptionalField(obj.dataset_composition.fasta?.sequencing_info?.assembly_details, isValidString, "FASTA assembly details must be a valid string.", "Optional - Description of assembly algorithms and parameters used in generating the genome assembly.");
        validateOptionalField(obj.dataset_composition.fasta?.sequencing_info?.error_rate, isValidFloat, "FASTA error rate must be a float value.", "Optional - Estimated error rate of the sequence data as a percentage (Float).");
        validateOptionalField(obj.dataset_composition.fasta?.further_source_details?.source_organism_info, isValidString, "FASTA source organism info must be a valid string.", "Optional - Detailed information on the source organisms used, including any unique genetic or phenotypic characteristics. This includes both naturally occurring and engineered traits.");
        validateOptionalField(obj.dataset_composition.fasta?.further_source_details?.NCBI_Taxonomy_ID, isValidString, "FASTA NCBI Taxonomy ID must be a valid string.", "Optional - The National Center for Biotechnology Information (NCBI) Taxonomy IDs, if applicable.");
        validateOptionalField(obj.dataset_composition.fasta?.further_source_details?.GenBank_Accession_Number, isValidString, "FASTA GenBank Accession Number must be a valid string.", "Optional - The GenBank accession number associated with the genomic sequences of the organisms, if applicable.");
        validateOptionalField(obj.dataset_composition.fasta?.further_source_details?.UniProt_ID, isValidString, "FASTA UniProt ID must be a valid string.", "Optional - The Universal Protein Resource (UniProt) IDs, if applicable.");
        validateOptionalField(obj.dataset_composition.genbank?.general_info?.specific_species, isValidString, "GenBank specific species must be a valid string.", "Optional - List of the specific species or strains represented in the GenBank files.");
        validateOptionalField(obj.dataset_composition.genbank?.sequencing_info?.coverage_depth, isValidInteger, "GenBank coverage depth must be an integer.", "Optional - Average depth of coverage, crucial for assessing data quality as an Integer.");
        validateOptionalField(obj.dataset_composition.genbank?.sequencing_info?.assembly_details, isValidString, "GenBank assembly details must be a valid string.", "Optional - Description of assembly algorithms and parameters used in generating the genome assembly.");
        validateOptionalField(obj.dataset_composition.genbank?.sequencing_info?.error_rate, isValidFloat, "GenBank error rate must be a float value.", "Optional - Estimated error rate of the sequence data as a percentage (Float).");
        validateOptionalField(obj.dataset_composition.genbank?.further_source_details?.source_organism_info, isValidString, "GenBank source organism info must be a valid string.", "Optional - Detailed information on the source organisms used, including any unique genetic or phenotypic characteristics. This includes both naturally occurring and engineered traits.");
        validateOptionalField(obj.dataset_composition.genbank?.further_source_details?.NCBI_Taxonomy_ID, isValidString, "GenBank NCBI Taxonomy ID must be a valid string.", "Optional - The National Center for Biotechnology Information (NCBI) Taxonomy IDs, if applicable.");
        validateOptionalField(obj.dataset_composition.genbank?.further_source_details?.GenBank_Accession_Number, isValidString, "GenBank GenBank Accession Number must be a valid string.", "Optional - The GenBank accession number associated with the genomic sequences of the organisms, if applicable.");
        validateOptionalField(obj.dataset_composition.genbank?.further_source_details?.UniProt_ID, isValidString, "GenBank UniProt ID must be a valid string.", "Optional - The Universal Protein Resource (UniProt) IDs, if applicable.");
        validateOptionalField(obj.dataset_composition.matlab?.toolboxes_used, isValidString, "Matlab toolboxes used must be a valid string.", "Optional - List of MATLAB toolboxes used in the data generation process.");
        validateOptionalField(obj.dataset_composition.text?.encoding, isValidString, "Text encoding must be a valid string.", "Optional - Character encoding used in text files ('UTF-8'/ 'UTF-16'/ 'UTF-32'/ 'ASCII'/ 'Other').");
        validateOptionalField(obj.dataset_composition.text?.language, isValidString, "Text language must be a valid string.", "Optional - Predominant language of text files (e.g., English, Spanish, Chinese).");
        validateOptionalField(obj.dataset_composition.pdf?.pdf_version, isValidString, "PDF version must be a valid string.", "Optional - Version of PDF format used (e.g., 1.4, 1.7).");
        validateOptionalField(obj.ethics?.data_collection_ethics, isValidString, "Data collection ethics must be a valid string.", "Optional - Information on ethical considerations related to data collection (e.g., consent, privacy).");
        validateOptionalField(obj.ethics?.data_usage_ethics, isValidString, "Data usage ethics must be a valid string.", "Optional - Information on ethical considerations related to data usage (e.g., intended use, potential misuse).");
        validateOptionalField(obj.ethics?.ethical_review_status, isValidString, "Ethical review status must be a valid string.", "Optional - Information on whether the dataset has undergone ethical review (e.g., approved, pending, not applicable)."); {% endcomment %}

        if (errors.length > 0) {
          errors.forEach((error) => {
            document.getElementById("dynamic-error-section").insertAdjacentHTML(
              "beforeend",
              `
              <div class="alert alert-danger alert-dismissible fade show" role="alert">
                <div class="text-center">
                  <i class="bi bi-exclamation-triangle-fill"></i>
                  <strong>${error}</strong>
                  <button type="button" class="btn-close" data-bs-dismiss="alert" aria-label="Close"></button>
                </div>
              </div>
            `
            );
          });

          // Clears input field if the value is incorrect.
          document.getElementById("json-file").value = "";
          return false;
        }
        return true;
      }
    </script>

    <!-- Upload Boxes Script -->
    <script>
      function updateFileInfo(input, infoId) {
        const infoDiv = document.getElementById(infoId);
        const parentDiv = input.closest(".drag-area");
        const defaultText = parentDiv.querySelector("p");
        const defaultIcon = parentDiv.querySelector("i");

        if (input.files && input.files[0]) {
          const file = input.files[0];

          // Determine the type of upload (folder or file) and update the display accordingly
          if (input.getAttribute("webkitdirectory") !== null) {
            // Folder upload
            let folderName = file.webkitRelativePath.split("/")[0];
            infoDiv.innerHTML = `<div class="file-display d-flex flex-column align-items-center justify-content-center">
                                <i class="bi bi-folder mb-2"></i>
                                <span>${folderName}</span>
                             </div>`;
          } else {
            // File upload
            infoDiv.innerHTML = `<div class="file-display d-flex flex-column align-items-center justify-content-center">
                                <i class="bi bi-file-earmark-text mb-2"></i>
                                <span>${file.name}</span>
                             </div>`;
          }

          // Style adjustments
          infoDiv.style.position = "absolute";
          infoDiv.style.top = "50%";
          infoDiv.style.left = "50%";
          infoDiv.style.transform = "translate(-50%, -50%)";
          infoDiv.style.textAlign = "center";

          // Hide the default text and icon
          defaultText.style.display = "none";
          defaultIcon.style.display = "none";

          parentDiv.classList.add("file-uploaded");
        } else {
          // Reset to default if no file or folder is selected
          defaultText.style.display = "";
          defaultIcon.style.display = "";
          infoDiv.innerHTML = "";
          parentDiv.classList.remove("file-uploaded");
        }
      }
    </script>

    <!-- Expandable List Script -->
    <script>
      function toggleVisibility(event, id) {
        event.stopPropagation(); // Prevent event bubbling to parent elements.
        var element = document.getElementById(id);
        var trigger = event.currentTarget;
        var triggerSymbol = trigger.getElementsByClassName('toggle-symbol')[0];

        if (element.style.display === 'block') {
          element.style.display = 'none';
          if (triggerSymbol) {
            triggerSymbol.textContent = '+';
          }
        } else {
          element.style.display = 'block';
          if (triggerSymbol) {
            triggerSymbol.textContent = '-';
          }
        }
      }
      </script>
    {% endblock %}
  </div>
</div>
